From 38014fe5857e37177f3887b9aafc70772ae3fb99 Mon Sep 17 00:00:00 2001
From: Suneel Garapati <sgarapati@caviumnetworks.com>
Date: Fri, 8 Sep 2017 16:52:58 -0700
Subject: [PATCH 0704/1239] drivers: rtc: fix ds1337 century base to 2000

Linux puts up century base as 2000 instead of 1900, resulting
in mismatch when either u-boot or linux sets date. Fix u-boot
to use century base as 2000 which clears the mismatch.

Signed-off-by: Suneel Garapati <sgarapati@caviumnetworks.com>
---
 drivers/rtc/ds1337.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/rtc/ds1337.c b/drivers/rtc/ds1337.c
index 8e220dfc54..903c78e86b 100644
--- a/drivers/rtc/ds1337.c
+++ b/drivers/rtc/ds1337.c
@@ -128,7 +128,7 @@ static int ds133x_rtc_get(struct udevice *dev, struct rtc_time *tmp)
 	tmp->tm_hour = bcd2bin (hour & 0x3F);
 	tmp->tm_mday = bcd2bin (mday & 0x3F);
 	tmp->tm_mon  = bcd2bin (mon_cent & 0x1F);
-	tmp->tm_year = bcd2bin (year) + ((mon_cent & 0x80) ? 2000 : 1900);
+	tmp->tm_year = bcd2bin (year) + ((mon_cent & 0x80) ? 2100 : 2000);
 	tmp->tm_wday = bcd2bin ((wday - 1) & 0x07);
 	tmp->tm_yday = 0;
 	tmp->tm_isdst= 0;
@@ -148,9 +148,13 @@ static int ds133x_rtc_set(struct udevice *dev, const struct rtc_time *tmp)
 		tmp->tm_year, tmp->tm_mon, tmp->tm_mday, tmp->tm_wday,
 		tmp->tm_hour, tmp->tm_min, tmp->tm_sec);
 
+	if (tmp->tm_year < 2000 || tmp->tm_year > 2199)
+		return -1;
+
 	dm_rtc_write (dev, RTC_YR_REG_ADDR, bin2bcd (tmp->tm_year % 100));
 
-	century = (tmp->tm_year >= 2000) ? 0x80 : 0;
+	/* Assume 20YY as base, to fix mismatch with linux */
+	century = (tmp->tm_year >= 2100) ? 0x80 : 0;
 	dm_rtc_write (dev, RTC_MON_REG_ADDR, bin2bcd (tmp->tm_mon) | century);
 
 	dm_rtc_write (dev, RTC_DAY_REG_ADDR, bin2bcd (tmp->tm_wday + 1));
-- 
2.29.0

